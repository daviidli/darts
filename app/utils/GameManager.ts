/* eslint-disable no-await-in-loop */
import DartBoard from './dartBoard';
import store from '../store/store';
import {
	setCurrentRound,
	setCurrentPlayer,
	initializedRound,
	setWinner,
	setWaiting,
	setMaxDarts,
	setMaxRounds,
	setCurrentDart,
	setTotals
} from '../actions/gameActions';

abstract class GameManager {
	private dartboard: DartBoard;
	private running: boolean;
	private missedDart: boolean;
	private dartsCleared: boolean;

	constructor() {
		const state = store.getState();

		if (!state.serialPort.length) {
			throw new Error('No serial port selected.');
		}

		this.dartboard = new DartBoard(state.serialPort);
		this.running = false;
		this.missedDart = false;
		this.dartsCleared = false;
	}

	protected abstract initialize(): void;
	protected abstract async turn(): Promise<void>;

	protected async run() {
		const {
			maxRounds,
			maxDarts,
			players,
			clearDartsWaitTime
		} = store.getState();

		const roundsArr = this.createStrArray(
			players.length,
			maxRounds,
			maxDarts
		);
		store.dispatch(initializedRound(roundsArr));

		this.running = true;
		for (let r = 0; r < maxRounds && this.running; r++) {
			store.dispatch(setCurrentRound(r));
			for (let p = 0; p < players.length && this.running; p++) {
				store.dispatch(setCurrentPlayer(p));
				await this.turn();
				await this.waitForClearingDarts(clearDartsWaitTime);
			}
		}
	}

	protected async getDart(): Promise<string> {
		let currentDart;
		while (true) {
			try {
				currentDart = await this.dartboard.getResults(1, 500);
			} catch (e) {
				// continue regardless of error
				// todo: show error notification
			}

			if (this.missedDart) {
				this.missedDart = false;
				return 'miss';
			}

			if (currentDart) {
				return currentDart[0];
			}
		}
	}

	public start() {
		this.initialize();
		this.run();
	}

	public stop() {
		this.running = false;
		this.dartboard.close();
		store.dispatch(setMaxDarts(0));
		store.dispatch(setMaxRounds(0));
		store.dispatch(setCurrentPlayer(0));
		store.dispatch(setCurrentDart(0));
		store.dispatch(setCurrentRound(0));
		store.dispatch(initializedRound([]));
		store.dispatch(setTotals([]));
	}

	public setWinner(index: number) {
		this.stop();
		store.dispatch(setWinner(index));
	}

	public miss() {
		this.missedDart = true;
	}

	public getRunning() {
		return this.running;
	}

	public clearedDarts() {
		this.dartsCleared = true;
	}

	private async waitForClearingDarts(timeout = 0) {
		this.dartsCleared = false;
		store.dispatch(setWaiting(true));
		let timeouts = 0;

		while (true) {
			await this.timeout(1000);
			timeouts += 1;
			if (timeouts === timeout) {
				store.dispatch(setWaiting(false));
				return;
			}
			if (this.dartsCleared) {
				this.dartsCleared = false;
				store.dispatch(setWaiting(false));
				return;
			}
		}
	}

	private timeout(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	private createStrArray(
		iMax: number,
		jMax: number,
		kMax: number
	): string[][][] {
		const initialRounds: string[][][] = [];
		for (let i = 0; i < iMax; i++) {
			const arr = [];
			for (let j = 0; j < jMax; j++) {
				arr.push(new Array(kMax).fill(''));
			}
			initialRounds.push(arr);
		}
		return initialRounds;
	}
}

export default GameManager;
