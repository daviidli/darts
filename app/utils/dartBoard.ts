import SerialPort from 'serialport';
import mappings from '../constants/mapping.json';

const Delimiter = require('@serialport/parser-readline');

class DartBoard {
	private parser: SerialPort;

	public constructor(port: string) {
		const p = new SerialPort(port, {
			baudRate: 9600
		});

		this.parser = p.pipe(new Delimiter({ delimiter: '\n' }));
	}

	public getResults(n: number, timeout = 10000): Promise<string[]> {
		this.parser.resume();
		return new Promise((resolve, reject) => {
			const results: string[] = [];

			let timeoutObj: NodeJS.Timeout;
			if (timeout !== -1) {
				timeoutObj = setTimeout(() => reject(), timeout);
			}

			this.parser.on('data', (data: string) => {
				results.push(data);

				if (results.length === n) {
					this.parser.pause();
					clearTimeout(timeoutObj);
					resolve(DartBoard.mapResults(results));
				}
			});
		});
	}

	private static mapResults(coords: string[]): string[] {
		return coords.map(coord => {
			const [x, y] = coord.split(',');
			return mappings.board[x][y];
		});
	}
}

export default DartBoard;
