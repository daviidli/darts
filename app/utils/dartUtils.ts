/* eslint-disable prefer-destructuring */
export const getValue = (dart: string): number => {
	switch (dart) {
		case 'miss':
		case '':
			return 0;
		case 'b':
			return 25;
		case 'b-d':
			return 50;
		default: {
			const arr = dart.split('-');
			if (arr.length === 1) {
				return parseInt(arr[0], 10);
			}

			const value = parseInt(arr[0], 10);
			if (arr[1] === 'd') {
				return value * 2;
			}

			return value * 3;
		}
	}
};

export const getPreviousRounds = (
	rounds: string[][],
	numRounds: number,
	currentRound: number
) => {
	if (currentRound < numRounds) {
		return rounds.slice(0, currentRound);
	}

	return rounds.slice(currentRound - numRounds, currentRound);
};

export type dartValue = {
	value: string;
	modifier: string;
};

export const mappingToString = (value: string) => {
	const ret: dartValue = {
		value: '',
		modifier: ''
	};
	switch (value) {
		case 'b':
			ret.value = 'Bull';
			return ret;
		case 'b-d':
			ret.value = 'Bull';
			ret.modifier = 'd';
			return ret;
		case 'miss':
			ret.value = 'Miss';
			return ret;
		default: {
			const arr = value.split('-');
			ret.value = arr[0];
			if (arr.length === 1) {
				return ret;
			}

			ret.modifier = arr[1];
			return ret;
		}
	}
};
