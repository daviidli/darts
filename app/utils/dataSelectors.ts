export const getRounds = (
	rounds: string[][],
	currentPlayer: number
): string[] => {
	return rounds.map(round => round[currentPlayer]);
};

export const test = () => {};
