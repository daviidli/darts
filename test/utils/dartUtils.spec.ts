import * as utils from '../../app/utils/dartUtils';

describe('getValue tests', () => {
	it('should get value for empty string', () => {
		expect(utils.getValue('')).toBe(0);
	});

	it('should get miss value', () => {
		expect(utils.getValue('miss')).toBe(0);
	});

	it('should get value for bullseye', () => {
		expect(utils.getValue('b')).toBe(25);
	});

	it('should get value for double bullseye', () => {
		expect(utils.getValue('b-d')).toBe(50);
	});

	it('should get value for regular value', () => {
		expect(utils.getValue('20')).toBe(20);
	});

	it('should get value for double value', () => {
		expect(utils.getValue('20-d')).toBe(40);
	});

	it('should get value for triple value', () => {
		expect(utils.getValue('20-t')).toBe(60);
	});
});

describe('getPreviousRounds tests', () => {
	it('should get previous rounds', () => {
		const rounds = [['1', '2', '3']];
		expect(utils.getPreviousRounds(rounds, 6, 1)).toStrictEqual(rounds);
	});

	it('should get previous rounds for a bigger example', () => {
		const rounds = [
			['1'],
			['2'],
			['3'],
			['4'],
			['5'],
			['6'],
			['7'],
			['8'],
			['9']
		];
		expect(utils.getPreviousRounds(rounds, 6, 8)).toMatchSnapshot();
	});
});

describe('mappingToString tests', () => {
	it('should get mapping for empty string', () => {
		expect(utils.mappingToString('')).toStrictEqual({
			value: 'Miss',
			modifier: ''
		});
	});

	it('should get mapping for miss', () => {
		expect(utils.mappingToString('miss')).toStrictEqual({
			value: 'Miss',
			modifier: ''
		});
	});

	it('should get mapping for bullseye', () => {
		expect(utils.mappingToString('b')).toStrictEqual({
			value: 'Bull',
			modifier: ''
		});
	});

	it('should get mapping for double bullseye', () => {
		expect(utils.mappingToString('b-d')).toStrictEqual({
			value: 'Bull',
			modifier: 'd'
		});
	});

	it('should get mapping for regular value', () => {
		expect(utils.mappingToString('20')).toStrictEqual({
			value: '20',
			modifier: ''
		});
	});

	it('should get mapping for double value', () => {
		expect(utils.mappingToString('20-d')).toStrictEqual({
			value: '20',
			modifier: 'd'
		});
	});

	it('should get mapping for triple value', () => {
		expect(utils.mappingToString('20-t')).toStrictEqual({
			value: '20',
			modifier: 't'
		});
	});
});
