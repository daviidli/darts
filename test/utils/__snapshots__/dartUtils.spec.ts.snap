// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPreviousRounds tests should get previous rounds for a bigger example 1`] = `
Array [
  Array [
    "3",
  ],
  Array [
    "4",
  ],
  Array [
    "5",
  ],
  Array [
    "6",
  ],
  Array [
    "7",
  ],
  Array [
    "8",
  ],
]
`;
